File: 백준_C++/5427.cpp
#include<bits/stdc++.h>
using namespace std;

int dx[4] = {0, 0, 1, -1};
int dy[4] = {1, -1, 0 ,0};
string board[1001];
queue<pair<int, int>> Q;
pair<int, int> start;
int T, w, h, fire[1001][1001], dist[1001][1001];

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);

    cin >> T;
    while(T--) {
        cin >> w >> h;
        for(int i = 0; i < h; i++) cin >> board[i];
        for (int i = 0; i < h; i++) {
            fill(fire[i], fire[i] + w, -1);
            fill(dist[i], dist[i] + w, -1);
        }

        for(int i = 0; i < h; i++) {
            for(int j = 0; j < w; j++) {
                if (board[i][j] == '*') {
                    Q.push({i, j});
                    fire[i][j] = 0;
                }
                if (board[i][j] == '@') {
                    start = {i, j};
                    dist[i][j] = 0;
                }
            }
        }
        
        while (!Q.empty()) {
            auto cur = Q.front(); Q.pop();
            for(int dir = 0; dir < 4; dir++) {
                int nx = dx[dir] + cur.first;
                int ny = dy[dir] + cur.second;
                if (nx < 0 || nx >= h || ny < 0 || ny >= w) continue;
                if (board[nx][ny] == '#' || fire[nx][ny] >= 0) continue;
                fire[nx][ny] = fire[cur.first][cur.second] + 1;
                Q.push({nx, ny});
            }
        }

        bool escape = false;
        Q.push(start);
        while (!Q.empty() && !escape) {
            auto cur = Q.front(); Q.pop();
            for(int dir = 0; dir < 4; dir++) {
                int nx = dx[dir] + cur.first;
                int ny = dy[dir] + cur.second;
                if (nx < 0 || nx >= h || ny < 0 || ny >= w) {
                    escape = true;
                    cout << dist[cur.first][cur.second] + 1 << "\n";
                    break;
                }
                if (board[nx][ny] == '#' || dist[nx][ny] >=0) continue;
                if (fire[nx][ny] != -1 && dist[cur.first][cur.second] + 1 >= fire[nx][ny]) continue;
                dist[nx][ny] = dist[cur.first][cur.second] + 1;
                Q.push({nx, ny});
            }
        }
        if (!escape) cout << "IMPOSSIBLE" << "\n";
    }
}

File: 백준_Java/5427.java
import java.io.*;
import java.util.*;

class Main {
    static final int[] dx = {0, 0, 1, -1};
    static final int[] dy = {1, -1, 0, 0};

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(br.readLine());

        while (t > 0) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            int m = Integer.parseInt(st.nextToken());
            int n = Integer.parseInt(st.nextToken());

            char[][] board = new char[n][m];
            for (int i = 0; i < n; i++) {
                String lines = br.readLine();
                for (int j = 0; j < m; j++) {
                    board[i][j] = lines.charAt(j);
                }
            }

            int result = solution(n, m, board);
            if (result == -1) {
                System.out.println("IMPOSSIBLE");
            } else {
                System.out.println(result);
            }
            t--;
        }
    }

    static int solution(int n, int m, char[][] board) {
        int[][] fire = new int[n][m];
        int[][] dist = new int[n][m];
        Queue<Pair> q = new LinkedList<>();
        Pair start = null;

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (board[i][j] == '*') {
                    fire[i][j] = 1;
                    q.add(new Pair(i, j));
                }

                if (board[i][j] == '@') {
                    dist[i][j] = 1;
                    start = new Pair(i, j);
                }
            }
        }

        while (!q.isEmpty()) {
            Pair cur = q.poll();
            for (int dir = 0; dir < 4; dir++) {
                int nx = cur.x + dx[dir];
                int ny = cur.y + dy[dir];

                if (nx < 0 || nx >= n || ny < 0 || ny >= m) continue;
                if (fire[nx][ny] > 0 || board[nx][ny] == '#') continue;

                fire[nx][ny] = fire[cur.x][cur.y] + 1;
                q.add(new Pair(nx, ny));
            }
        }

        q.add(start);
        while (!q.isEmpty()) {
            Pair cur = q.poll();
            for (int dir = 0; dir < 4; dir++) {
                int nx = cur.x + dx[dir];
                int ny = cur.y + dy[dir];
                if (nx < 0 || nx >= n || ny < 0 || ny >= m) {
                    return dist[cur.x][cur.y];
                }
                if (dist[nx][ny] > 0 || board[nx][ny] == '#') continue;
                if (fire[nx][ny] != 0 && dist[cur.x][cur.y] + 1 >= fire[nx][ny]) continue;
                dist[nx][ny] = dist[cur.x][cur.y] + 1;
                q.add(new Pair(nx, ny));
            }
        }
        return -1;
    }

    static class Pair{
        int x;
        int y;
        public Pair(int x, int y) {
            this.x = x;
            this.y = y;
        }
    }
}